File: ios/Game6561/ViewModel/GameViewModel.swift
import Foundation
import Combine
import SwiftUI

@MainActor
class GameViewModel: ObservableObject {

    @Published var state = GameState()
    @Published var statistics = GameStatistics(gamesPlayed: 0, gamesWon: 0, totalScore: 0, bestScore: 0, totalMoves: 0, maxCombo: 0)

    @Published var showTutorial = false
    @Published var showSettings = false
    @Published var showStatistics = false
    @Published var showWinDialog = false
    @Published var showGameOverDialog = false

    @Published var isDarkTheme = false
    @Published var isSoundEnabled = true

    private let prefs = GamePreferences()
    let soundManager = SoundManager()
    private var timer: Timer?

    init() {
        soundManager.setEnabled(prefs.isSoundEnabled)
        isSoundEnabled = prefs.isSoundEnabled
        isDarkTheme = prefs.isDarkTheme
        statistics = prefs.statistics

        if let saved = prefs.loadGameState() {
            state = saved
            state.gamesPlayed = statistics.gamesPlayed
            state.gamesWon = statistics.gamesWon
            state.best = statistics.bestScore
            startTimer()
        } else {
            newGame()
        }

        showTutorial = !prefs.hasSeenTutorial
    }

    func move(direction: Direction) {
        let newState = GameEngine.move(state, direction: direction)
        guard newState != state else {
            soundManager.play(.invalid)
            return
        }

        let wasWon = state.gameWon
        let merged = newState.totalMerges > state.totalMerges
        state = newState
        prefs.saveGameState(newState)

        if newState.gameOver {
            stopTimer()
            statistics = prefs.recordGame(won: newState.gameWon, finalScore: newState.score, finalMoves: newState.moves, finalMaxCombo: newState.maxCombo)
            showGameOverDialog = true
            soundManager.play(.gameOver)
        } else {
            soundManager.play(.move)
            if merged {
                soundManager.play(newState.combo > 1 ? .combo : .merge)
            }
        }

        if newState.gameWon && !wasWon {
            soundManager.play(.win)
            showWinDialog = true
        }
    }

    func undo() {
        guard let result = GameEngine.undo(state) else { return }
        state = result
        prefs.saveGameState(result)
        soundManager.play(.undo)
    }

    func newGame() {
        stopTimer()
        state = GameEngine.createInitialState()
        state.gamesPlayed = statistics.gamesPlayed
        state.gamesWon = statistics.gamesWon
        state.best = statistics.bestScore
        showWinDialog = false
        showGameOverDialog = false
        startTimer()
        soundManager.play(.start)
        prefs.clearGameState()
    }

    func keepPlaying() {
        showWinDialog = false
        state.gameWon = false
    }

    func toggleTheme() {
        isDarkTheme.toggle()
        prefs.isDarkTheme = isDarkTheme
    }

    func toggleSound() {
        isSoundEnabled.toggle()
        soundManager.setEnabled(isSoundEnabled)
        prefs.isSoundEnabled = isSoundEnabled
    }

    func dismissTutorial() {
        showTutorial = false
        prefs.hasSeenTutorial = true
    }

    func resetTutorial() {
        prefs.resetTutorial()
        showTutorial = true
    }

    func refreshStatistics() {
        statistics = prefs.statistics
    }

    // MARK: - Timer

    private func startTimer() {
        stopTimer()
        timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            Task { @MainActor in
                guard let self = self, !self.state.gameOver else { return }
                self.state.gameTimer += 1
            }
        }
    }

    private func stopTimer() {
        timer?.invalidate()
        timer = nil
    }

    deinit {
        stopTimer()
    }
}

func formatTime(_ seconds: Int) -> String {
    let mins = seconds / 60
    let secs = seconds % 60
    return String(format: "%02d:%02d", mins, secs)
}

File: android/app/src/main/java/com/game6561/app/GameViewModel.kt
package com.game6561.app

import android.app.Application
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.game6561.app.data.GamePreferences
import com.game6561.app.game.Direction
import com.game6561.app.game.GameEngine
import com.game6561.app.game.GameState
import com.game6561.app.game.GameStatistics
import com.game6561.app.sound.SoundManager
import com.game6561.app.sound.SoundType
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

class GameViewModel(application: Application) : AndroidViewModel(application) {

    private val prefs = GamePreferences(application)
    val soundManager = SoundManager(application)

    var state by mutableStateOf(GameState())
        private set

    var statistics by mutableStateOf(GameStatistics(0, 0, 0, 0, 0, 0))
        private set

    var showTutorial by mutableStateOf(false)
        private set

    var showSettings by mutableStateOf(false)
    var showStatistics by mutableStateOf(false)

    var showWinDialog by mutableStateOf(false)
        private set

    var showGameOverDialog by mutableStateOf(false)
        private set

    var isDarkTheme by mutableStateOf(false)
        private set

    var isSoundEnabled by mutableStateOf(true)
        private set

    private var timerJob: Job? = null

    init {
        soundManager.setEnabled(prefs.isSoundEnabled())
        isSoundEnabled = prefs.isSoundEnabled()
        isDarkTheme = prefs.getTheme() == "dark"
        statistics = prefs.getStatistics()

        val saved = prefs.loadGameState()
        if (saved != null) {
            state = saved.copy(
                gamesPlayed = statistics.gamesPlayed,
                gamesWon = statistics.gamesWon,
                best = statistics.bestScore
            )
            if (!state.gameOver && !state.gameWon) startTimer()
        } else {
            newGame()
        }

        showTutorial = !prefs.hasSeenTutorial()
    }

    fun move(direction: Direction) {
        val newState = GameEngine.move(state, direction)
        if (newState !== state) {
            val wasWon = state.gameWon
            val merged = newState.totalMerges > state.totalMerges
            state = newState
            prefs.saveGameState(newState)

            if (newState.gameOver) {
                stopTimer()
                statistics = prefs.recordGame(
                    won = newState.gameWon,
                    finalScore = newState.score,
                    finalMoves = newState.moves,
                    maxCombo = newState.maxCombo
                )
                showGameOverDialog = true
                soundManager.play(SoundType.GAME_OVER)
            } else {
                soundManager.play(SoundType.MOVE)
                if (merged) {
                    soundManager.play(if (newState.combo > 1) SoundType.COMBO else SoundType.MERGE)
                }
            }

            if (newState.gameWon && !wasWon) {
                soundManager.play(SoundType.WIN)
                showWinDialog = true
            }
        } else {
            soundManager.play(SoundType.INVALID)
        }
    }

    fun undo() {
        val result = GameEngine.undo(state)
        if (result != null) {
            state = result
            prefs.saveGameState(result)
            soundManager.play(SoundType.UNDO)
        }
    }

    fun newGame() {
        stopTimer()
        state = GameEngine.createInitialState().copy(
            gamesPlayed = state.gamesPlayed,
            gamesWon = state.gamesWon,
            best = statistics.bestScore
        )
        showWinDialog = false
        showGameOverDialog = false
        startTimer()
        soundManager.play(SoundType.START)
        prefs.clearGameState()
    }

    /** App 进入后台：暂停计时器 */
    fun onAppBackgrounded() {
        stopTimer()
    }

    /** App 回到前台：恢复计时器 */
    fun onAppForegrounded() {
        if (!state.gameOver) {
            startTimer()
        }
    }

    fun keepPlaying() {
        showWinDialog = false
        state = state.copy(gameWon = false)
    }

    fun toggleTheme() {
        isDarkTheme = !isDarkTheme
        prefs.setTheme(if (isDarkTheme) "dark" else "light")
    }

    fun toggleSound() {
        isSoundEnabled = !isSoundEnabled
        soundManager.setEnabled(isSoundEnabled)
        prefs.setSoundEnabled(isSoundEnabled)
    }

    fun dismissTutorial() {
        showTutorial = false
        prefs.markTutorialSeen()
    }

    fun resetTutorial() {
        prefs.resetTutorial()
        showTutorial = true
    }

    fun dismissWin() {
        showWinDialog = false
        keepPlaying()
    }

    fun dismissGameOver() {
        showGameOverDialog = false
    }

    override fun onCleared() {
        super.onCleared()
        stopTimer()
        soundManager.release()
    }

    private fun startTimer() {
        timerJob?.cancel()
        timerJob = viewModelScope.launch {
            while (isActive) {
                delay(1000)
                if (!state.gameOver) {
                    state = state.copy(gameTimer = state.gameTimer + 1)
                }
            }
        }
    }

    private fun stopTimer() {
        timerJob?.cancel()
        timerJob = null
    }

    fun updateStatisticsDisplay() {
        statistics = prefs.getStatistics()
    }
}
